### PR TITLE
avoid concurrent sub/unsub on same channel

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>
issue: #26451

### CODE REVIEW

File: internal/querynodev2/services.go [hunk 1]
```diff
@@ -253,6 +253,13 @@ func (node *QueryNode) WatchDmChannels(ctx context.Context, req *querypb.WatchDm
 	}
 	defer node.subscribingChannels.Remove(channel.GetChannelName())
 
+	// to avoid concurrent watch/unwatch
+	if node.unsubscribingChannels.Contain(channel.GetChannelName()) {
```
Comment (yah01) [inline internal/querynodev2/services.go:257]: how does this work if the unsubscribing is coming but has not modified the `unsubscribingChannels`?

File: internal/querynodev2/services.go [hunk 2]
```diff
@@ -375,6 +382,7 @@ func (node *QueryNode) UnsubDmChannel(ctx context.Context, req *querypb.UnsubDmC
 		return status, nil
 	}
 
+	node.unsubscribingChannels.Insert(req.GetChannelName())
 	delegator, ok := node.delegators.GetAndRemove(req.GetChannelName())
```
Comment (yah01) [inline internal/querynodev2/services.go:386]: Use `defer` to make it always work even new `return` inserted

**Review by yah01 — CHANGES_REQUESTED** [review]

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**